### PR TITLE
Revise cohort structure inside community

### LINKED ISSUE #336: Revise cohort structure inside Community
At the moment, the `Community` object `stem_traits` and `stem_allometry` attributes that are instances of `StemTraits` and `StemAllometry` classes. And then it has cohort data, which is a just a dictionary:

https://github.com/ImperialCollegeLondon/pyrealm/blob/30da5273150af8add4ec5367779f441b0f4e374d/pyrealm/demography/community.py#L355-L364

1. It would be much more consistent to have this as a dataclass - and it would make downstream code more consistent in using dot notation.

2. It would also make sense to be able to apply the ABC classes to add shared functionality to demography classes being worked on in #314 to the cohort data (pandas output, mechanism for adding and deleting 'rows' across cohorts).

So, this should be a dataclass:

```python
@dataclass
class Cohorts(PandasExporter):
    cohort_dbh_values: NDArray[np.float32]
    cohort_n_individuals: NDArray[np.int_]
    cohort_pft_names: NDArray[np.str_]
```

I don't know if we make this just internal to `Community` or whether we make a `Cohorts` instance part of the signature for creating a community, replacing the three `NDArray` arguments. Seems cleaner to use it in the signature.

### PR DESCRIPTION
**Pull request by davidorme**
# Description

This PR:

* replaces the `community.cohort_data: dict[str, NDArray]` dictionary of arrays with a new dataclass `Cohorts`.
* updates the test suite to add some simple testing of the `Cohorts` constructor
* updates the rest of the test suite and the code to the new structure. 

The main reasons are:

* more consistent attribute access within model rather than a mix of dataclass dot notation and dictionary keys.
* ability to extend the functionality of the Cohorts dataclass with shared functionality (see #338).

Fixes #336

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pyrealm/demography/community.py [hunk 1]
```diff
@@ -133,6 +135,39 @@
...
+    """
+
+    dbh_values: NDArray[np.float64]
+    n_individuals: NDArray[np.int_]
+    pft_names: NDArray[np.str_]
+
+    def __post_init__(self) -> None:
+        """Validation of cohorts data."""
+
+        # TODO - validation - maybe make this optional to reduce workload within
```
[the remark this answers is not in the document]
Comment (davidorme) [reply pyrealm/demography/community.py:153]: Bah - forgot to add the existing issue number #317 ! It's a reminder of a new place to look within handling that issue.

File: tests/unit/demography/test_community.py [hunk 2]
```diff
@@ -52,119 +52,145 @@ def check_expected(community, expected):
...
+            "Cohort data not passed as numpy arrays",
             id="lists_not_arrays",
         ),
         pytest.param(
             {
-                "cell_id": 1,
-                "cell_area": 100,
-                "cohort_pft_names": np.array(["broadleaf", "conifer"]),
-                "cohort_n_individuals": np.array([6, 1]),
-                "cohort_dbh_values": np.array([0.2, 0.5, 0.9]),
```
[the remark this answers is not in the document]
Comment (davidorme) [reply tests/unit/demography/test_community.py:83]: Yes - that's an issue that is now handled when a `Cohorts` instance is created (arrays not the same length) so this test case has just moved up into the `test_Cohorts` parameterisation.

[the remark this answers is not in the document]
Comment (MarionBWeinzierl) [reply tests/unit/demography/test_community.py:83]: OK, that makes snese

**Review by MarionBWeinzierl — APPROVED** [review]
The changes look ok to me (with the explanations provided), and this seems a cleaner way to implement it.

**Pull request merged**